hoot-core/src/main/cpp/hoot/core/algorithms/WayJoiner.cpp

```diff
@@ -0,0 +1,321 @@
+/*
+ * This file is part of Hootenanny.
+ *
+ * Hootenanny is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2018 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "WayJoiner.h"
+
+//  Hoot
+#include <hoot/core/conflate/NodeToWayMap.h>
+#include <hoot/core/index/OsmMapIndex.h>
+#include <hoot/core/ops/RecursiveElementRemover.h>
+#include <hoot/core/ops/ReplaceElementOp.h>
+#include <hoot/core/schema/OsmSchema.h>
+#include <hoot/core/schema/TagMergerFactory.h>
+
+#include <unordered_set>
+#include <vector>
+
+using namespace std;
+
+namespace hoot
+{
+
+WayJoiner::WayJoiner(const OsmMapPtr &map)
+  : _map(map),
+    _leavePid(true)
+{
+}
+
+void WayJoiner::join(const OsmMapPtr &map)
+{
+  WayJoiner joiner(map);
+  joiner.join();
+}
+
+void WayJoiner::join()
+{
+  //  Join back any ways with parent ids
+  joinParentChild();
+  //  Join any siblings that have the same parent id but the parent isn't connected
+  joinSiblings();
+  //  Rejoin any ways that are now connected to their parents
+  joinParentChild();
+  //  Run one last join on ways that share a node and have a parent id
+  joinAtNode();
+  //  Clear out any remaining unjoined parent ids
+  resetParents();
+}
+
+void WayJoiner::resetParents()
+{
+  if (_leavePid)
+    return;
+  WayMap ways = _map->getWays();
+  //  Find all ways that have a split parent id and reset them
+  for (WayMap::const_iterator it = ways.begin(); it != ways.end(); ++it)
+  {
+    WayPtr way = it->second;
+    if (way->hasPid())
+      way->resetPid();
+  }
+}
+
+void WayJoiner::joinParentChild()
+{
+  WayMap ways = _map->getWays();
+  vector<long> ids;
+  //  Find all ways that have a split parent id
+  for (WayMap::const_iterator it = ways.begin(); it != ways.end(); ++it)
+  {
+    WayPtr way = it->second;
+    if (way->hasPid())
+      ids.push_back(way->getId());
+  }
+  if (ids.size() == 0)
```

Comment:
I think this is unnecessary - if ids.size() is zero, the following sort should do nothing, and the following for loop should do nothing. I would vote for a cleaner control flow, and remove this return statement.

Reply:
Good point.